### PR TITLE
add VOneNets using new StochasticModelCommitment 

### PR DESCRIPTION
**Pull request by tiagogmarques**
- StochasticModelCommitment and VOneCORnetCommitment classes to extract activations multiple times (one per trial)
- StochasticModelPool with base_models that should use the StochasticModel

### CODE REVIEW

File: candidate_models/base_models/__init__.py
Comment (mschrimpf) [inline candidate_models/base_models/__init__.py:509]: I think the CORnet-S version is missing from here (I see the changes further down, but no model)

**Comment by mschrimpf** [comment]
the unit tests seem to be failing (click the X next to Travis CI for log) @tiagogmarques

File: candidate_models/base_models/__init__.py
Comment (tiagogmarques) [reply candidate_models/base_models/__init__.py:509]: That's because the CORnet-S version is defined only at the brain-model level and is part of the VOneCORnetBrainPool in candidate_models/model_commitments/vonecornets

**Comment by tiagogmarques** [comment]
> the unit tests seem to be failing (click the X next to Travis CI for log) @tiagogmarques

From what I could read in the Brain-Score Jenkons CI log (could not find any log in the Travis CI), the error is because the brainscore repo has not been updated with my PR. The MLBrainPool initializer should accept a new boolean parameter "stochastic" to generate a brain pool either with the ModelCommitment or the StochasticModelCommitment classes.

File: candidate_models/base_models/__init__.py
Comment (mschrimpf) [reply candidate_models/base_models/__init__.py:509]: ah got it

**Comment by mschrimpf** [comment]
for the travis log, you need to click on "Details", and then on "The build". 

I'm a bit lost why there needs to be a `stochastic` flag? Aren't stochastic models always stochastic? Sorry, I should have asked that in the other PR (can also discuss there). The `ModelCommitment` is supposed to be the "standard choice" for models, so if we think they should be stochastic then they should always be stochastic

**Comment by tiagogmarques** [comment]
> for the travis log, you need to click on "Details", and then on "The build".
> 
> I'm a bit lost why there needs to be a `stochastic` flag? Aren't stochastic models always stochastic? Sorry, I should have asked that in the other PR (can also discuss there). The `ModelCommitment` is supposed to be the "standard choice" for models, so if we think they should be stochastic then they should always be stochastic

I had 2 options, either create a MLStochasticBrainPool class, which would be the same as the MLBrainPool but using the StochasticModelCommitment to build the brain model from the corresponding base model, or change the MLBrainPool to accept a parameter that controls which brain model class it uses (StochasticModelCommitment or ModelCommitment). I opted for the second. If you prefer the first one I can implement it that way.

**Comment by mschrimpf** [comment]
Yes I think for now I would create `MLStochasticBrainPool` class in `candidate_models` like you described in option 1. (Trying to avoid multiple case distinctions within the "standard setup" as much as possible to keep it straight-forward for external users.)

**Comment by mschrimpf** [comment]
re-opening to trigger jenkins tests

**Comment by mschrimpf** [comment]
re-triggering again

**Comment by mschrimpf** [comment]
Looks like a bunch of tests failed, did these work locally?
```
FAILED tests/test___init__.py::TestPreselectedLayer::test_alexnet_conv2_V4 - ...
FAILED tests/test___init__.py::TestPreselectedLayer::test_alexnet_conv5_V4 - ...
FAILED tests/test___init__.py::TestPreselectedLayer::test_alexnet_conv5_IT - ...
FAILED tests/test___init__.py::TestPreselectedLayer::test_repeat_same_result
FAILED tests/test___init__.py::TestPreselectedLayer::test_newmodel_pytorch - ...
FAILED tests/test___init__.py::TestBrainTranslated::test_MajajHong2015ITpls[alexnet-0.59033-True]
FAILED tests/test___init__.py::TestBrainTranslated::test_MajajHong2015ITpls[resnet18-supervised-0.596013-False]
FAILED tests/test___init__.py::TestBrainTranslated::test_MajajHong2015ITpls[resnet18-local_aggregation-0.60152-False]
FAILED tests/test___init__.py::TestBrainTranslated::test_MajajHong2015ITpls[resnet18-autoencoder-0.373528-False]
FAILED tests/test___init__.py::TestBrainTranslated::test_candidate_Kar2019OST[CORnet-S-0.240888]
FAILED tests/test___init__.py::TestBrainTranslated::test_candidate_Kar2019OST[CORnet-R2-0.230859]
FAILED tests/test___init__.py::TestBrainTranslated::test_candidate_Kar2019OST[alexnet-nan]
FAILED tests/test_activations.py::TestActivations::test_model[resnet18-simclr]
FAILED tests/base_models/test___init__.py::test_run_logits[vgg-16] - Exceptio...
FAILED tests/base_models/test___init__.py::test_run_logits[vggface] - Attribu...
FAILED tests/base_models/test___init__.py::test_run_logits[densenet-121] - At...
FAILED tests/base_models/test___init__.py::test_run_logits[densenet-169] - At...
FAILED tests/base_models/test___init__.py::test_run_logits[densenet-201] - At...
FAILED tests/base_models/test___init__.py::test_run_logits[convrnn_224] - Mod...
FAILED tests/base_models/test___init__.py::test_run_logits[resnet18-supervised]
FAILED tests/base_models/test___init__.py::test_run_logits[resnet18-local_aggregation]
FAILED tests/base_models/test___init__.py::test_run_logits[resnet18-instance_recognition]
FAILED tests/base_models/test___init__.py::test_run_logits[resnet18-autoencoder]
FAILED tests/base_models/test___init__.py::test_run_logits[resnet18-contrastive_predictive]
FAILED tests/base_models/test___init__.py::test_run_logits[resnet18-colorization]
FAILED tests/base_models/test___init__.py::test_run_logits[resnet18-relative_position]
FAILED tests/base_models/test___init__.py::test_run_logits[resnet18-depth_prediction]
FAILED tests/base_models/test___init__.py::test_run_logits[prednet] - TypeErr...
FAILED tests/base_models/test___init__.py::test_run_logits[resnet18-simclr]
FAILED tests/base_models/test___init__.py::test_run_logits[resnet18-deepcluster]
FAILED tests/base_models/test___init__.py::test_run_logits[resnet18-contrastive_multiview]
FAILED tests/base_models/test___init__.py::test_identifier[densenet-169-densenet169]
```

**Comment by mschrimpf** [comment]
Some of the errors seem to stem from keras such as `TypeError: 'dict' object is not reversible` or `AttributeError: 'str' object has no attribute 'decode'` when running `original_keras_version = f.attrs['keras_version'].decode('utf8')` or `URL fetch failure on https://github.com/fchollet/deep-learning-models/releases/download/v0.1/vgg16_weights_tf_dim_ordering_tf_kernels.h5 : None -- retrieval incomplete: got only 28997608 out of 553467096 bytes`

**Comment by mschrimpf** [comment]
I don't think the failing tests are due to changes from this PR. Let's add unit tests for VOneNet to ensure that works and I will try to figure out why the other tests fail.

File: candidate_models/model_commitments/vonecornets.py
Comment (mschrimpf) [inline candidate_models/model_commitments/vonecornets.py:31]: dead code

File: candidate_models/base_models/__init__.py
Comment (mschrimpf) [inline candidate_models/base_models/__init__.py:273]: if I understand right, VOne{CORnet,Resnet} are now exclusively defined in the brain model pool no?

File: candidate_models/model_commitments/vonecornets.py
Comment (mschrimpf) [inline candidate_models/model_commitments/vonecornets.py:13]: is setting this field needed?

**Review by mschrimpf — APPROVED** [review]
two clarification questions, but looks good to go otherwise as long as the new unit tests pass!

File: candidate_models/model_commitments/vonecornets.py
Comment (tiagogmarques) [reply candidate_models/model_commitments/vonecornets.py:13]: Yes. I also made a change in the CORnetCommitment to set that field. That's because the naming of the layers differs from the typical CORnet and so it cannot extract the region recorded from the activations layer in the look_at_cached method

File: candidate_models/base_models/__init__.py
Comment (tiagogmarques) [reply candidate_models/base_models/__init__.py:273]: there are 3 new models on the brain model pool: 'VOneCORnet-S', 'voneresnet-50' and 'voneresnet-50-robust. I don't understand what you mean by exclusively defined. The base model versions of the 'voneresnet-50' and 'voneresnet-50-robust' can also be accessed.

Comment (tiagogmarques) [reply candidate_models/base_models/__init__.py:273]: I removed them from the base model pool in my commitment and added them to a StochasticModelPool. They should not be in the base model bool.

Comment (mschrimpf) [reply candidate_models/base_models/__init__.py:273]: I believe you removed them from the list (https://github.com/brain-score/candidate_models/pull/70/files#diff-30b56a69b0b5a99bcaf36381fc10683e55aff0b7a42516cd0ee8f6d1ff9c12a0L391-L393) but the function definitions are still there (https://github.com/brain-score/candidate_models/pull/70/files#diff-30b56a69b0b5a99bcaf36381fc10683e55aff0b7a42516cd0ee8f6d1ff9c12a0R251-R273, lines 251 to 274 in `base_models/__init__.py`)

**Comment by mschrimpf** [comment]
openmind ran into a node failure, re-starting

**Comment by mschrimpf** [comment]
openmind failed again, re-triggering

**Comment by mschrimpf** [comment]
Tests that fail (and are relevant to this PR):
```
FAILED tests/test___init__.py::TestPreselectedLayer::test_newmodel_pytorch - ...
FAILED tests/test___init__.py::TestBrainTranslated::test_candidate_Kar2019OST[CORnet-S-0.240888]
FAILED tests/test___init__.py::TestBrainTranslated::test_candidate_Kar2019OST[CORnet-R2-0.230859]
FAILED tests/test___init__.py::TestBrainTranslated::test_candidate_Kar2019OST[VOneCORnet-S-0.23107]
FAILED tests/test___init__.py::TestBrainTranslated::test_Rajalingham2018i2n[VOneCORnet-S-0.356]
FAILED tests/test___init__.py::TestBrainTranslated::test_Rajalingham2018i2n[voneresnet-50-0.371]
FAILED tests/test___init__.py::TestBrainTranslated::test_Rajalingham2018i2n[voneresnet-50-robust-0.386]
```
You might have to merge `master` into your branch, I think that will address the OST benchmark

File: candidate_models/base_models/__init__.py
Comment (mschrimpf) [reply candidate_models/base_models/__init__.py:273]: bump @tiagogmarques

File: candidate_models/base_models/__init__.py
Comment (mschrimpf) [inline candidate_models/base_models/__init__.py:258]: I'm a bit confused by this -- since the VOneNets defined here no longer use anything stochastic, is it even necessary to have a separate pool for them? (https://github.com/brain-score/candidate_models/pull/70/files#diff-30b56a69b0b5a99bcaf36381fc10683e55aff0b7a42516cd0ee8f6d1ff9c12a0R498)

**Comment by mschrimpf** [comment]
for some reason the OST tests still fail:
```
FAILED tests/test___init__.py::TestBrainTranslated::test_candidate_Kar2019OST[CORnet-S-expected_score0]
FAILED tests/test___init__.py::TestBrainTranslated::test_candidate_Kar2019OST[CORnet-R2-expected_score1]
FAILED tests/test___init__.py::TestBrainTranslated::test_candidate_Kar2019OST[VOneCORnet-S-expected_score3]
```
e.g.
```
    def pipe(self, source_assembly, target_assembly):
        # check only for equal values, alignment is given by metadata
>       assert sorted(source_assembly[self._split_coord].values) == sorted(target_assembly[self._split_coord].values)
E       AssertionError
```
as well as one of the behavioral tests for VOneResNet
```
FAILED tests/test___init__.py::TestBrainTranslated::test_Rajalingham2018i2n[voneresnet-50-robust-expected_score4]
```

**Comment by tiagogmarques** [comment]
I had not merged the setup.py from the master which may have caused the OST issue?
The problem with the voneresnet-50-robust was resolved.

---
Tiago Marques PhD
Postdoctoral Researcher
MIT Department of Brain and Cognitive Sciences
PhRMA Foundation Fellow




On Apr 15, 2021, at 11:23 AM, Martin Schrimpf ***@***.******@***.***>> wrote:



for some reason the OST tests still fail:

FAILED tests/test___init__.py::TestBrainTranslated::test_candidate_Kar2019OST[CORnet-S-expected_score0]
FAILED tests/test___init__.py::TestBrainTranslated::test_candidate_Kar2019OST[CORnet-R2-expected_score1]
FAILED tests/test___init__.py::TestBrainTranslated::test_candidate_Kar2019OST[VOneCORnet-S-expected_score3]


e.g.

    def pipe(self, source_assembly, target_assembly):
        # check only for equal values, alignment is given by metadata
>       assert sorted(source_assembly[self._split_coord].values) == sorted(target_assembly[self._split_coord].values)
E       AssertionError


as well as one of the behavioral tests for VOneResNet

FAILED tests/test___init__.py::TestBrainTranslated::test_Rajalingham2018i2n[voneresnet-50-robust-expected_score4]

**Comment by mschrimpf** [comment]
cool that solved most of them. The remaining ones are 
```
FAILED tests/test___init__.py::TestBrainTranslated::test_candidate_Kar2019OST[CORnet-S-expected_score0]
FAILED tests/test___init__.py::TestBrainTranslated::test_candidate_Kar2019OST[CORnet-R2-expected_score1]
FAILED tests/test___init__.py::TestBrainTranslated::test_candidate_Kar2019OST[VOneCORnet-S-expected_score3]
```
because of `assert sorted(source_assembly[self._split_coord].values) == sorted(target_assembly[self._split_coord].values)`. 
I frankly don't understand what has changed to cause this, do you have an idea?

**Comment by mschrimpf** [comment]
^ bumping this

**Comment by mschrimpf** [comment]
https://github.com/brain-score/brain-score/pull/268 should fix the OST failures we are seeing, will re-trigger after that one has been merged

**Comment by mschrimpf** [comment]
re-trigger

**Comment by mschrimpf** [comment]
there also seems to be a new xarray issue that affects the OST tests:
```
    def test_candidate_Kar2019OST(self, model_identifier, expected_score):
        model = brain_translated_pool[model_identifier]
        score = score_model(model_identifier=model_identifier, model=model, benchmark_identifier='dicarlo.Kar2019-ost')
>       if not np.isnan(expected_score):
E       TypeError: operand 'ApproxScalar' does not support ufuncs (__array_ufunc__=None)
```

**Comment by mschrimpf** [comment]
```
    def look_at(self, stimuli, number_of_trials=1):
        stimuli_identifier = stimuli.identifier
        for trial_number in range(number_of_trials):
>           stimuli.identifier = stimuli_identifier + '-trial' + f'{trial_number:03d}'
E           TypeError: unsupported operand type(s) for +: 'NoneType' and 'str'

candidate_models/model_commitments/vonecornets.py:23: TypeError
```
You probably need to check for `None` in the stimulus_set identifier

**Comment by mschrimpf** [comment]
None of the remaining failures imo have anything to do with this PR 
```
FAILED tests/test___init__.py::TestPreselectedLayer::test_newmodel_pytorch - ...
FAILED tests/test___init__.py::TestBrainTranslated::test_MajajHong2015ITpls[resnet18-local_aggregation-expected_score2-False]
FAILED tests/test___init__.py::TestBrainTranslated::test_MajajHong2015ITpls[resnet18-autoencoder-expected_score3-False]
FAILED tests/test_activations.py::TestActivations::test_model[resnet18-simclr]
FAILED tests/base_models/test___init__.py::test_run_logits[vggface] - Attribu...
FAILED tests/base_models/test___init__.py::test_run_logits[densenet-121] - At...
FAILED tests/base_models/test___init__.py::test_run_logits[densenet-169] - At...
FAILED tests/base_models/test___init__.py::test_run_logits[densenet-201] - At...
FAILED tests/base_models/test___init__.py::test_run_logits[resnext101_32x32d_wsl]
FAILED tests/base_models/test___init__.py::test_run_logits[convrnn_224] - Mod...
FAILED tests/base_models/test___init__.py::test_run_logits[resnet18-supervised]
FAILED tests/base_models/test___init__.py::test_run_logits[resnet18-local_aggregation]
FAILED tests/base_models/test___init__.py::test_run_logits[resnet18-instance_recognition]
FAILED tests/base_models/test___init__.py::test_run_logits[resnet18-autoencoder]
FAILED tests/base_models/test___init__.py::test_run_logits[resnet18-contrastive_predictive]
FAILED tests/base_models/test___init__.py::test_run_logits[resnet18-colorization]
FAILED tests/base_models/test___init__.py::test_run_logits[resnet18-relative_position]
FAILED tests/base_models/test___init__.py::test_run_logits[resnet18-depth_prediction]
FAILED tests/base_models/test___init__.py::test_run_logits[prednet] - TypeErr...
FAILED tests/base_models/test___init__.py::test_run_logits[resnet18-simclr]
FAILED tests/base_models/test___init__.py::test_run_logits[resnet18-deepcluster]
FAILED tests/base_models/test___init__.py::test_run_logits[resnet18-contrastive_multiview]
FAILED tests/base_models/test___init__.py::test_identifier[densenet-169-densenet169]
```

Merging

**Pull request merged**
